File: apps/player-linux/src/renderer/countdown-bar-resolver.ts
/**
 * Countdown Bar schedule resolution: which configured bar — if any — should be
 * on screen right now. Kept as a tiny global beside the countdown display
 * because renderer scripts run directly in the sandboxed page without module
 * imports, and shared so the renderer surface and its tests evaluate the exact
 * same weekly, one-time, DST, completion, and priority behavior.
 */
/**
 * One entry of the manifest's plugin array as it arrives, before any resolver
 * has claimed it. A resolver narrows by `type` and `version` and then reads its
 * own configuration shape, so an entry belonging to another plugin — or to a
 * plugin this Player predates — travels through untouched.
 */
interface TilecastManifestPluginEntry {
  id: string;
  type: string;
  version: number;
  config: unknown;
}

interface TilecastCountdownBarPlugin {
  id: string;
  type: string;
  version: number;
  config: {
    message: string;
    scheduleType: "weekly" | "one_time";
    targetTime?: string | null;
    daysOfWeek?: number[];
    oneTimeAt?: string | null;
    timezone: string;
    leadTimeSeconds: number;
    completionText?: string;
    showConfetti?: boolean;
    displayMode: "overlay" | "push";
    heightPx: number;
    progressFill?: "none" | "drain" | null;
    contentPadding?: number | null;
    textScale?: number | null;
    urgencyEnabled?: boolean;
    startingSoonSeconds?: number | null;
    urgentSeconds?: number | null;
    pulseSeconds?: number | null;
    priority: number;
  };
}

interface TilecastActiveCountdownBar {
  id: string;
  message: string;
  value: string;
  displayMode: "overlay" | "push";
  heightPx: number;
  priority: number;
  targetAt: string;
  completed: boolean;
  showBar: boolean;
  showConfetti: boolean;
  urgencyStage: "normal" | "starting_soon" | "urgent" | "final";
  urgencyLabel: string;
  pulse: boolean;
  /**
   * Share of the lead window still to run, 1 when the bar first appears and 0 at
   * the target. Always 0 while completion text shows. `null` when the instance
   * asks for no fill, so the surface can leave the background alone rather than
   * paint a full-width tint.
   */
  remainingFraction: number | null;
  /** Gutter on each side, as a percentage of the bar width. */
  contentPadding: number;
  /** Type size in pixels, already scaled. */
  fontSizePx: number;
}

interface TilecastCountdownBarResolver {
  resolve(
    plugins: TilecastManifestPluginEntry[] | null | undefined,
    localNow: Date,
    clockOffsetMs?: number,
  ): TilecastActiveCountdownBar | null;
}

const tilecastCountdownBar: TilecastCountdownBarResolver = (() => {
  const COMPLETION_DISPLAY_MS = 60_000;
  const CONFETTI_DISPLAY_MS = 12_000;
  const WEEKDAYS = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
  // One formatter per zone: a bar re-resolves every second on hardware where
  // rebuilding Intl formatters is the most expensive part of the tick.
  const formatters = new Map<string, Intl.DateTimeFormat>();

  function formatter(timezone: string): Intl.DateTimeFormat {
    let cached = formatters.get(timezone);
    if (!cached) {
      cached = new Intl.DateTimeFormat("en-US", {
        timeZone: timezone,
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit",
        hourCycle: "h23",
        weekday: "short",
      });
      formatters.set(timezone, cached);
    }
    return cached;
  }

  function zonedParts(at: Date, timezone: string) {
    const parts = formatter(timezone).formatToParts(at);
    const value = (type: Intl.DateTimeFormatPartTypes) =>
      parts.find((part) => part.type === type)?.value ?? "";
    return {
      year: Number(value("year")),
      month: Number(value("month")),
      day: Number(value("day")),
      hour: Number(value("hour")),
      minute: Number(value("minute")),
      second: Number(value("second")),
      weekday: WEEKDAYS.indexOf(value("weekday")),
    };
  }

  /** Convert a calendar wall time in an IANA zone to an absolute instant. */
  function zonedInstant(
    year: number,
    month: number,
    day: number,
    hour: number,
    minute: number,
    second: number,
    timezone: string,
  ): number {
    const desired = Date.UTC(year, month - 1, day, hour, minute, second);
    let guess = desired;
    // Two passes cover ordinary offsets and DST boundary corrections without a
    // heavy timezone dependency. Intl remains the source of timezone truth.
    for (let pass = 0; pass < 3; pass += 1) {
      const actual = zonedParts(new Date(guess), timezone);
      const represented = Date.UTC(
        actual.year,
        actual.month - 1,
        actual.day,
        actual.hour,
        actual.minute,
        actual.second,
      );
      const correction = desired - represented;
      guess += correction;
      if (correction === 0) break;
    }
    return guess;
  }

  /** Manifests may omit these, and a hand-edited one may be out of range. */
  function clampNumber(
    value: number | null | undefined,
    fallback: number,
    lowest: number,
    highest: number,
  ): number {
    if (typeof value !== "number" || !Number.isFinite(value)) return fallback;
    return Math.min(highest, Math.max(lowest, value));
  }

  /**
   * Type size follows the bar height, clamped the way the stylesheet used to
   * clamp it, then scaled. The scale multiplies the clamped result so a bar can
   * exceed the original 72px ceiling deliberately rather than by accident.
   */
  function countdownFontSizePx(heightPx: number, scalePercent: number): number {
    const base = Math.min(72, Math.max(22, heightPx * 0.42));
    return Math.round(base * (scalePercent / 100) * 100) / 100;
  }

  function candidateTargets(
    plugin: TilecastCountdownBarPlugin,
    now: Date,
  ): number[] {
    const config = plugin.config;
    if (config.scheduleType === "one_time") {
      const instant = Date.parse(config.oneTimeAt ?? "");
      return Number.isFinite(instant) ? [instant] : [];
    }
    const match = /^(\d{2}):(\d{2})(?::(\d{2}))?$/.exec(
      config.targetTime ?? "",
    );
    if (!match) return [];
    const hour = Number(match[1]);
    const minute = Number(match[2]);
    const second = Number(match[3] ?? 0);
    if (hour > 23 || minute > 59 || second > 59) return [];
    const current = zonedParts(now, config.timezone);
    const dateAnchor = Date.UTC(current.year, current.month - 1, current.day);
    const days = new Set(config.daysOfWeek ?? []);
    const targets: number[] = [];
    // A lead window may begin up to 30 days before its target. Looking in both
    // directions also retains completion text around the preceding occurrence.
    for (let offset = -31; offset <= 31; offset += 1) {
      const date = new Date(dateAnchor + offset * 86_400_000);
      if (!days.has(date.getUTCDay())) continue;
      targets.push(
        zonedInstant(
          date.getUTCFullYear(),
          date.getUTCMonth() + 1,
          date.getUTCDate(),
          hour,
          minute,
          second,
          config.timezone,
        ),
      );
    }
    return targets;
  }

  return Object.freeze({
    resolve(
      plugins: TilecastManifestPluginEntry[] | null | undefined,
      localNow: Date,
      clockOffsetMs = 0,
    ): TilecastActiveCountdownBar | null {
      const now = localNow.getTime() + clockOffsetMs;
      const active: TilecastActiveCountdownBar[] = [];
      for (const entry of plugins ?? []) {
        if (entry.type !== "countdown_bar" || entry.version !== 1) continue;
        const plugin = entry as TilecastCountdownBarPlugin;
        for (const target of candidateTargets(plugin, new Date(now))) {
          const remaining = target - now;
          const completionText = plugin.config.completionText?.trim() ?? "";
          const completionVisible =
            remaining <= 0 &&
            remaining >= -COMPLETION_DISPLAY_MS &&
            completionText.length > 0;
          const confettiVisible =
            remaining <= 0 &&
            remaining >= -CONFETTI_DISPLAY_MS &&
            plugin.config.showConfetti === true;
          if (
            remaining > plugin.config.leadTimeSeconds * 1_000 ||
            (remaining <= 0 && !completionVisible && !confettiVisible)
          ) {
            continue;
          }
          // A non-positive lead window would divide by zero; treat it as spent
          // rather than emitting NaN into a CSS width.
          const leadMs = plugin.config.leadTimeSeconds * 1_000;
          const fraction =
            leadMs > 0 ? Math.min(1, Math.max(0, remaining / leadMs)) : 0;
          const startingSoonMs =
            clampNumber(plugin.config.startingSoonSeconds, 300, 2, 86_400) *
            1_000;
          const urgentMs =
            clampNumber(plugin.config.urgentSeconds, 60, 2, 3_600) * 1_000;
          const pulseMs =
            clampNumber(plugin.config.pulseSeconds, 10, 1, 60) * 1_000;
          const urgencyStage =
            plugin.config.urgencyEnabled !== true || remaining <= 0
              ? "normal"
              : remaining <= pulseMs
                ? "final"
                : remaining <= urgentMs
                  ? "urgent"
                  : remaining <= startingSoonMs
                    ? "starting_soon"
                    : "normal";
          const finalScale = urgencyStage === "final" ? 1.25 : 1;
          const baseHeight = Math.min(
            Math.max(plugin.config.heightPx, 40),
            320,
          );
          active.push({
            id: plugin.id,
            // Completion text is the whole ending message, not a replacement
            // for only the numeric value. Keeping the normal prefix here would
            // produce phrases such as "Lunch ends in Lunch is over".
            message: remaining <= 0 ? "" : plugin.config.message,
            value: completionVisible
              ? completionText
              : remaining > 0
                ? tilecastCountdownDisplay.compact(remaining)
                : "",
            displayMode: plugin.config.displayMode,
            heightPx: Math.round(baseHeight * finalScale),
            priority: plugin.config.priority,
            targetAt: new Date(target).toISOString(),
            completed: remaining <= 0,
            showBar: remaining > 0 || completionVisible,
            showConfetti: confettiVisible,
            urgencyStage,
            urgencyLabel:
              urgencyStage === "starting_soon"
                ? "Starting soon"
                : urgencyStage === "urgent" || urgencyStage === "final"
                  ? "Urgent"
                  : "",
            pulse: urgencyStage === "final",
            remainingFraction:
              plugin.config.progressFill === "drain" ? fraction : null,
            contentPadding: clampNumber(plugin.config.contentPadding, 4, 0, 40),
            fontSizePx:
              countdownFontSizePx(
                plugin.config.heightPx,
                clampNumber(plugin.config.textScale, 100, 25, 500),
              ) * finalScale,
          });
        }
      }
      active.sort(
        (left, right) =>
          right.priority - left.priority ||
          Date.parse(left.targetAt) - Date.parse(right.targetAt) ||
          left.id.localeCompare(right.id),
      );
      return active[0] ?? null;
    },
  });
})();

// Exposed for unit tests only. In the player this is a plain global shared
// between the renderer scripts, which have no module loader.
(
  globalThis as typeof globalThis & {
    tilecastCountdownBar: TilecastCountdownBarResolver;
  }
).tilecastCountdownBar = tilecastCountdownBar;

File: apps/player-android/app/src/main/java/org/tilecast/player/content/CountdownBarResolver.kt
package org.tilecast.player.content

import org.tilecast.player.network.ManifestPlugin
import java.time.DayOfWeek
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.OffsetDateTime
import java.time.ZoneId
import kotlin.math.roundToInt

/**
 * Countdown Bar schedule resolution. This mirrors the Linux renderer's
 * countdown-bar-resolver: the same weekly, one-time, DST, completion, priority,
 * and fill rules, so a bar targeted at both platforms behaves identically.
 * Recurrence is evaluated locally from the cached manifest, so an offline player
 * keeps showing and hiding bars on schedule.
 */
internal data class ActiveCountdownBar(
    val id: String,
    val message: String,
    val value: String,
    val displayMode: String,
    val heightPx: Int,
    val priority: Int,
    val targetAt: Instant,
    val completed: Boolean,
    val showBar: Boolean,
    val showConfetti: Boolean,
    val urgencyStage: String,
    val urgencyLabel: String,
    val pulse: Boolean,
    /**
     * Share of the lead window still to run: 1 when the bar appears, 0 at the
     * target, and 0 while completion text shows. Null when the instance asked
     * for no fill.
     */
    val remainingFraction: Float?,
    /** Gutter on each side, as a percentage of the bar width. */
    val contentPadding: Int,
    /** Type size in scale-independent pixels, already scaled. */
    val fontSizeSp: Float,
)

private val CompletionDisplay: Duration = Duration.ofMinutes(1)
private val ConfettiDisplay: Duration = Duration.ofSeconds(12)

internal fun resolveCountdownBar(
    plugins: List<ManifestPlugin>,
    now: Instant,
    clockOffsetSeconds: Long? = null,
    clockOffsetMillis: Long? = null,
): ActiveCountdownBar? {
    val at = now.plusMillis(clockOffsetMillis ?: (clockOffsetSeconds ?: 0) * 1_000L)
    val active = mutableListOf<ActiveCountdownBar>()
    for (plugin in plugins) {
        if (plugin.type != "countdown_bar" || plugin.version != 1) continue
        val config = plugin.config
        val leadMs = config.leadTimeSeconds * 1_000L
        for (target in countdownBarTargets(plugin, at)) {
            val remainingMs = Duration.between(at, target).toMillis()
            val completionText = config.completionText.trim()
            val completionVisible =
                remainingMs <= 0 &&
                    remainingMs >= -CompletionDisplay.toMillis() &&
                    completionText.isNotEmpty()
            val confettiVisible =
                remainingMs <= 0 &&
                    remainingMs >= -ConfettiDisplay.toMillis() &&
                    config.showConfetti
            if (remainingMs > leadMs || (remainingMs <= 0 && !completionVisible && !confettiVisible)) continue
            val fraction = if (leadMs > 0) (remainingMs.toFloat() / leadMs).coerceIn(0f, 1f) else 0f
            val urgencyStage =
                when {
                    !config.urgencyEnabled || remainingMs <= 0 -> "normal"
                    remainingMs <= config.pulseSeconds.coerceIn(1, 60) * 1_000L -> "final"
                    remainingMs <= config.urgentSeconds.coerceIn(2, 3_600) * 1_000L -> "urgent"
                    remainingMs <= config.startingSoonSeconds.coerceIn(2, 86_400) * 1_000L -> "starting_soon"
                    else -> "normal"
                }
            val finalScale = if (urgencyStage == "final") 1.25f else 1f
            val baseHeight = config.heightPx.coerceIn(40, 320)
            active +=
                ActiveCountdownBar(
                    id = plugin.id,
                    // Completion text is the whole ending message. Retaining
                    // the normal prefix would render phrases such as
                    // "Lunch ends in Lunch is over".
                    message = if (remainingMs <= 0) "" else config.message,
                    value =
                        when {
                            completionVisible -> completionText
                            remainingMs > 0 -> compactCountdown(remainingMs)
                            else -> ""
                        },
                    displayMode = config.displayMode,
                    heightPx = (baseHeight * finalScale).roundToInt(),
                    priority = config.priority,
                    targetAt = target,
                    completed = remainingMs <= 0,
                    showBar = remainingMs > 0 || completionVisible,
                    showConfetti = confettiVisible,
                    urgencyStage = urgencyStage,
                    urgencyLabel =
                        when (urgencyStage) {
                            "starting_soon" -> "Starting soon"
                            "urgent", "final" -> "Urgent"
                            else -> ""
                        },
                    pulse = urgencyStage == "final",
                    remainingFraction = if (config.progressFill == "drain") fraction else null,
                    contentPadding = config.contentPadding.coerceIn(0, 40),
                    fontSizeSp = countdownBarFontSize(config.heightPx, config.textScale) * finalScale,
                )
        }
    }
    return active
        .sortedWith(
            compareByDescending<ActiveCountdownBar> { it.priority }
                .thenBy { it.targetAt }
                .thenBy { it.id },
        )
        .firstOrNull()
}

/**
 * Candidate targets around now. A lead window may open up to 30 days before its
 * target, and looking backwards as well keeps completion text alive just after
 * the preceding occurrence.
 */
private fun countdownBarTargets(plugin: ManifestPlugin, now: Instant): List<Instant> {
    val config = plugin.config
    if (config.scheduleType == "one_time") {
        val parsed =
            config.oneTimeAt?.let { value ->
                runCatching { Instant.parse(value) }.getOrNull()
                    ?: runCatching { OffsetDateTime.parse(value).toInstant() }.getOrNull()
            }
        return listOfNotNull(parsed)
    }
    val time = parseTargetTime(config.targetTime) ?: return emptyList()
    if (config.daysOfWeek.isEmpty()) return emptyList()
    val zone = runCatching { ZoneId.of(config.timezone) }.getOrDefault(ZoneId.of("UTC"))
    val days = config.daysOfWeek.mapNotNull(::isoDayOfWeek).toSet()
    if (days.isEmpty()) return emptyList()
    val today = now.atZone(zone).toLocalDate()
    val targets = mutableListOf<Instant>()
    for (offset in -31L..31L) {
        val date: LocalDate = today.plusDays(offset)
        if (date.dayOfWeek !in days) continue
        // ZonedDateTime resolves a wall time that a DST jump skipped or repeated,
        // which is what keeps a 12:00 bar at local noon across the boundary.
        targets += date.atTime(time).atZone(zone).toInstant()
    }
    return targets
}

/** Manifest days are 0=Sunday through 6=Saturday. */
private fun isoDayOfWeek(day: Int): DayOfWeek? =
    when (day) {
        0 -> DayOfWeek.SUNDAY
        1 -> DayOfWeek.MONDAY
        2 -> DayOfWeek.TUESDAY
        3 -> DayOfWeek.WEDNESDAY
        4 -> DayOfWeek.THURSDAY
        5 -> DayOfWeek.FRIDAY
        6 -> DayOfWeek.SATURDAY
        else -> null
    }

/** Accepts the HH:MM the API publishes and the HH:MM:SS Postgres can render. */
private fun parseTargetTime(value: String?): LocalTime? {
    val match = Regex("^(\\d{2}):(\\d{2})(?::(\\d{2}))?$").matchEntire(value ?: "") ?: return null
    val hour = match.groupValues[1].toIntOrNull() ?: return null
    val minute = match.groupValues[2].toIntOrNull() ?: return null
    val second = match.groupValues[3].takeIf { it.isNotEmpty() }?.toIntOrNull() ?: 0
    if (hour > 23 || minute > 59 || second > 59) return null
    return LocalTime.of(hour, minute, second)
}

/**
 * Type size follows the bar height, clamped as the Linux stylesheet clamps it,
 * then scaled. The scale multiplies the clamped result so a bar can exceed the
 * unscaled 72 ceiling deliberately rather than by accident.
 */
internal fun countdownBarFontSize(heightPx: Int, textScale: Int): Float {
    val base = (heightPx * 0.42f).coerceIn(22f, 72f)
    return base * (textScale.coerceIn(25, 500) / 100f)
}

/**
 * The same compact vocabulary the Linux renderer and Studio use, so one bar does
 * not read "5m 4s" on one platform and "5:04" on another.
 */
internal fun compactCountdown(remainingMilliseconds: Long): String {
    if (remainingMilliseconds <= 0) return "Now"
    val totalSeconds = remainingMilliseconds / 1_000
    val days = totalSeconds / 86_400
    val hours = (totalSeconds % 86_400) / 3_600
    val minutes = (totalSeconds % 3_600) / 60
    val seconds = totalSeconds % 60
    return when {
        days > 0 -> "${days}d ${hours}h"
        hours > 0 -> "${hours}h ${minutes}m"
        minutes > 0 -> "${minutes}m ${seconds}s"
        else -> "${seconds}s"
    }
}
